### ISSUE TITLE
Recursively defined datatype can lead to NPE in Java

### ISSUE THREAD

**Issue by robin-aws** [dafny-lang/dafny#5089]
### Dafny version

4.4.0 (present since at least 4.2 though)

### Code to produce this issue

```dafny
datatype Error =
  | CollectionOfErrors(list: seq<Error>, nameonly message: string)

method Main() {
  var e: Error;
}
```

### Command to run and resulting output

```code
% dafny run --target:java Scratch.dfy

Dafny program verifier finished with 0 verified, 0 errors
Exception in thread "main" java.lang.ExceptionInInitializerError
        at _System.__default.Main(__default.java:12)
        at _System.__default.__Main(__default.java:15)
        at Scratch.lambda$main$0(Scratch.java:9)
        at dafny.Helpers.withHaltHandling(Helpers.java:345)
        at Scratch.main(Scratch.java:9)
Caused by: java.lang.NullPointerException: Cannot invoke "dafny.TypeDescriptor.newArray(int)" because "<parameter1>" is null
        at dafny.ArrayDafnySequence.empty(DafnySequence.java:506)
        at dafny.DafnySequence.empty(DafnySequence.java:68)
        at _System.Error.<clinit>(Error.java:44)
        ... 5 more
```

### What happened?

The issue is this line in the translated Java code in Error.java:

```java
private static final Error theDefault = _System.Error.create(dafny.DafnySequence.<Error> empty(Error._typeDescriptor()), dafny.DafnySequence.<dafny.CodePoint> empty(dafny.TypeDescriptor.UNICODE_CHAR));
```

`Error._typeDescriptor()` depends on another static field (`_TYPE`), and Java semantics do not guarantee this field is initialized before `theDefault` is initialized.

Can be worked around by adding more data constructors before the recursively defined one, so that a different one is picked as the default value.

(I'm sure this was logged already but I can't find the issue I recall seeing before)

### What type of operating system are you experiencing the problem on?

Mac

**Comment by ssomayyajula** [comment]
This was logged #3766 which I (thought I) fixed in #4694; in fact, I'm unable to reproduce this in `master`. The Java spec guarantees that static fields are initialized in the order they appear in source, so if we look at `Error.java`...

```java
  private static final dafny.TypeDescriptor<Error> _TYPE = dafny.TypeDescriptor.<Error>referenceWithInitializer(Error.class, () -> Error.Default());
  public static dafny.TypeDescriptor<Error> _typeDescriptor() {
    return (dafny.TypeDescriptor<Error>) (dafny.TypeDescriptor<?>) _TYPE;
  }

  private static final Error theDefault = _System.Error.create(dafny.DafnySequence.<Error> empty(Error._typeDescriptor()), dafny.DafnySequence.<dafny.CodePoint> empty(dafny.TypeDescriptor.UNICODE_CHAR));
```

...it should all work out.

**Comment by ssomayyajula** [comment]
Oh, duh, 4.4.0 was released before #4694 was merged. Closing!
